src/tests.rs

```diff
@@ -0,0 +1,298 @@
+    assert_eq!(resampled, expected);
+}
+
+#[test]
+fn test_resampling_with_slow_data() {
+    let start = DateTime::from_timestamp(0, 0).unwrap();
+    let mut resampler: Resampler<f64, TestSample> =
+        Resampler::new(TimeDelta::seconds(1), ResamplingFunction::Average, 1, start);
+    let offset = TimeDelta::milliseconds(500);
+    let step = TimeDelta::seconds(2);
+    let data = vec![
+        TestSample::new(start + step * 1 - offset, Some(3.0)),
+        TestSample::new(start + step * 2 - offset, Some(4.0)),
+        TestSample::new(start + step * 3 - offset, Some(5.0)),
+        TestSample::new(start + step * 4 - offset, Some(6.0)),
+    ];
+
+    resampler.extend(data);
+
+    let expected = vec![
+        TestSample::new(DateTime::from_timestamp(1, 0).unwrap(), None),
+        TestSample::new(DateTime::from_timestamp(2, 0).unwrap(), Some(3.0)),
+        TestSample::new(DateTime::from_timestamp(3, 0).unwrap(), Some(3.0)),
+        TestSample::new(DateTime::from_timestamp(4, 0).unwrap(), Some(4.0)),
+        TestSample::new(DateTime::from_timestamp(5, 0).unwrap(), Some(4.0)),
+        TestSample::new(DateTime::from_timestamp(6, 0).unwrap(), Some(5.0)),
+        TestSample::new(DateTime::from_timestamp(7, 0).unwrap(), Some(5.0)),
+        TestSample::new(DateTime::from_timestamp(8, 0).unwrap(), Some(6.0)),
+    ];
+
+    let resampled = resampler.resample(start + step * 4);
+    assert_eq!(resampled, expected);
+}
+
+#[test]
+fn test_resampling_with_gap_early_end_date() {
+    let start = DateTime::from_timestamp(0, 0).unwrap();
+    let mut resampler: Resampler<f64, TestSample> =
+        Resampler::new(TimeDelta::seconds(5), ResamplingFunction::Average, 0, start);
+    let step = TimeDelta::seconds(1);
+    let data = vec![
+        TestSample::new(start, Some(1.0)),
+        TestSample::new(start + step, Some(2.0)),
+        TestSample::new(start + step * 3, Some(4.0)),
+        TestSample::new(start + step * 4, Some(5.0)),
+        TestSample::new(start + step * 16, Some(6.0)),
+        TestSample::new(start + step * 19, Some(10.0)),
+    ];
+
+    resampler.extend(data);
+
+    let expected = vec![
+        TestSample::new(DateTime::from_timestamp(5, 0).unwrap(), Some(3.0)),
+        TestSample::new(DateTime::from_timestamp(10, 0).unwrap(), None),
+    ];
+
+    let resampled = resampler.resample(start + step * 10);
+    assert_eq!(resampled, expected);
+
+    let expected2 = vec![
+        TestSample::new(DateTime::from_timestamp(15, 0).unwrap(), None),
+        TestSample::new(DateTime::from_timestamp(20, 0).unwrap(), Some(8.0)),
+    ];
+
+    let resampled2 = resampler.resample(start + step * 20);
+    assert_eq!(resampled2, expected2);
+}
+
+#[test]
+fn test_empty_buffer() {
+    let start = DateTime::from_timestamp(0, 0).unwrap();
+    let mut resampler: Resampler<f64, TestSample> =
+        Resampler::new(TimeDelta::seconds(5), ResamplingFunction::Average, 0, start);
```

Comment:
Check: Is 0 different from 1? If so, would 3 mean using the data from 4 intervals?

Reply:
Yes, that is correct! We resample, then we empty the buffer until `current - interval * max_age`. If `max_age` is `0`, no old values remain and in the next interval we only have values from this interval. If `max_age` is `1`, we have the last interval and the new one. Here are tests that show the behavior: https://github.com/frequenz-floss/frequenz-resampling-rs/pull/1/files#diff-a0262fea108af44a98cc49e5eb72641963dd816513f2d0a22eb738fcfed55ebeR138

Reply:
To me this feels not intuitive, because a `max_age` of 1 takes data from 2 sample intervals. It could change its behavior to shift by 1, or the name could change (e.g. `reuse_previous_intervals`, but probably there is a better name).

Reply:
So what are we doing about this?

Reply:
okay, sounds reasonable

Reply:
However... It does not make sense to set it to 0. Also regarding the way it is implemented now, I suggest validating the input and not allowing 0.

Reply:
See new test with required behavior: https://github.com/frequenz-floss/frequenz-resampling-rs/pull/1/files#diff-a0262fea108af44a98cc49e5eb72641963dd816513f2d0a22eb738fcfed55ebeR174

Reply:
OK, it seems this changed from the last review, so probably the readme needs updating too?

Reply:
>So if max_age is 0, the no values are used, and if it is 1, just the data from this window are used, etc.

So 0 should be forbidden then right? Would make sense to me, that `max_age` gives the number of sampling intervals with data, i.e. 1 contains only data from 1 sampling interval. 

W.r.t. naming, I understand to keep the desire to keep the existing name, but tbh I think it's not clear at all. To me `max_age` would be a `timedelta`.

Reply:
Yes, it wouldn't make sense to use `0`.

Reply:
I think this is the best solution

Reply:
The `max_age` as a multiplier has one advantage though and that's the adaptability to a slower input interval. If you set `interval` to 1s, but the data comes in every 2s, then the interval adapts to this frequency (to avoid empty resamples). This interval gets multiplied with `max_age`. 

With `max_age` being an interval itself, we can only assume that an adaptation to `1 x input_interval` is fine.

Reply:
Yeah, good point. That too.

Reply:
In my opinion it should be a multiple of the interval. My comment was just about the name of the parameter.

Reply:
People are still typing `max_age`, but wasn't it already renamed to `max_age_in_invervals`?

Reply:
So, what do we do? 

1. Keeping `max_age_in_invervals` (my preference)
2. Renaming it to `reuse_previous_intervals` or something similar
3. Making it an interval

Reply:
Also voting for 1.

Reply:
:one: